tamr_client/operation.py

```diff
@@ -0,0 +1,188 @@
+    )
+
+
+def succeeded(operation: Operation) -> bool:
+    """Convenience method for checking if operation was successful.
+    """
+    return operation.status is not None and operation.status["state"] == "SUCCEEDED"
+
+
+def _from_response(instance: Instance, response: requests.Response) -> Operation:
+    """
+    Handle idiosyncrasies in constructing Operations from Tamr responses.
+    When a Tamr API call would start an operation, but all results that would be
+    produced by that operation are already up-to-date, Tamr returns `HTTP 204 No Content`
+
+    To make it easy for client code to handle these API responses without checking
+    the response code, this method will either construct an Operation, or a
+    dummy `NoOp` operation representing the 204 Success response.
+
+    Args:
+        response: HTTP Response from the request that started the operation.
+    """
+    if response.status_code == 204:
+        # Operation was successful, but the response contains no content.
+        # Create a dummy operation to represent this.
+        _never = "0000-00-00T00:00:00.000Z"
+        _description = """Tamr returned HTTP 204 for this operation, indicating that all
+            results that would be produced by the operation are already up-to-date."""
+        resource_json = {
+            "id": "-1",
+            "type": "NOOP",
+            "description": _description,
+            "status": {
+                "state": "SUCCEEDED",
+                "startTime": _never,
+                "endTime": _never,
+                "message": "",
+            },
+            "created": {"username": "", "time": _never, "version": "-1"},
+            "lastModified": {"username": "", "time": _never, "version": "-1"},
+            "relativeId": "operations/-1",
+        }
+    else:
+        resource_json = response.json()
+    _id = resource_json["id"]
+    _url = URL(instance=instance, path=f"operations/{_id}")
+    return _from_json(_url, resource_json)
+
+
+def _from_url(session: Session, url: URL) -> Operation:
+    """Get operation by URL
+
+    Fetches operation from Tamr server
+
+    Args:
+        url: Operation URL
+
+    Raises:
+        OperationNotFound: If no operation could be found at the specified URL.
+            Corresponds to a 404 HTTP error.
+        requests.HTTPError: If any other HTTP error is encountered.
+    """
+    r = session.get(str(url))
+    if r.status_code == 404:
+        raise NotFound(str(url))
+    data = response.successful(r).json()
+    return _from_json(url, data)
+
+
+def _from_json(url: URL, data: JsonDict):
+    """Make operation from JSON data (deserialize)
+
+    Args:
+        url: Operation URL
+        data: Operation JSON data from Tamr server
+    """
+    cp = deepcopy(data)
+    return Operation(
+        url, type=cp["type"], status=cp.get("status"), description=cp.get("description")
+    )
+
+
+def _apply_options(
```

Comment:
another option we have is to make all functions that return an operation synchronous. Then, if people want asynchronous versions, we can create separate asynchronous functions.

E.g.

```python
def refresh(...) -> Operation:
   ...
   op = _refresh_async(...)
   return tc.operation.wait(op)

def _refresh_async(...) -> Operation:
    ...
```

----

^this design removes the need for a function like `_apply_options` since the synchronous version cannot have options applied at all, and the asynchronous option let's you apply your operation options by calling `tc.operation.wait(...)` yourself:

```python
op = _refresh_async(...)
tc.operation.wait(op, **my_options)
```

Reply:
I would prefer synchronous operations by default, but perhaps I'm biased by the current operation of the client. Perhaps this would warrant an RFC somewhere down the line?

Reply:
We already did investigation into whether sync or async should be the default in the early TUC days, and the response was overwhelmingly "sync". I don't think this warrants an RFC unless we think something has changed.